source/load-balancers/load-balancers.rst

```diff
@@ -0,0 +1,564 @@
+the session and the operation MUST perform normal server selection.*
+
+In the case of a network errors in transactions, the driver MUST implement the following cases:
+
+- When a network error occurs on a :code:`commitTransaction`, the connection MUST be closed
+  and the operation retried once with a newly checked out connection from the pool.
+- When a network error occurs on an :code:`abortTransaction`, the connection MUST be closed
+  and the operation retried once with a newly checked out connection from the pool. Any
+  subsequent errors from the abortTransaction command MUST be ignored.
+- When a network error occurs in any other operation inside a transaction, the connection
+  MUST be closed and unpinned and returned to the pool. The driver MUST add a
+  :code:`"TransientTransactionError"` label to the error.
+
+The driver connection pool MUST track the purpose for which connections are checked out
+in the following 3 categories:
+
+- Connections checked out for cursors
+- Connections checked out for transactions
+- Connections checked out for operations not falling under the previous 2 categories
+
+When the connection pool’s :code:`maxPoolSize` is reached and the pool times out waiting
+for a new connection the :code:`WaitQueueTimeoutError` MUST include a new detailed message,
+as well as the published :code:`ConnectionCheckOutFailedEvent`: *“Timeout waiting for
+connection from the connection pool. maxPoolSize: n, connections in use by cursors: n,
+connections in use by transactions: n, connections in use by other operations: n”*.
+
+
+**Error Handling**
+------------------
+
+When the driver is in load balanced mode and encounters any error described here as a state
+change error, the driver MUST NOT make any changes to the :code:`TopologyDescription` or the
+:code:`ServerDescription` of the load balancer (i.e. it MUST NOT mark the load balancer as Unknown).
+If the error requires the connection pool to be cleared, the driver MUST only clear connections
+with the same :code:`serviceId` as the connection which errored. The connection pool MUST emit a
+:code:`PoolClearedEvent` event with the :code:`serviceId` of the connection that errored.
+The :code:`PoolClearedEvent` MUST have a new field of :code:`serviceId` with type :code:`ObjectId`.
+
+If there is a network error or timeout on the connection before the initial handshake completes,
+the driver MUST NOT make any changes to the :code:`TopologyDescription` or mark the server as Unknown.
+The driver MUST also close the associated connection. If the network error occurred on the handshake’s
+:code:`hello` command, the connection pool MUST NOT be cleared. If the network error occurred after
+the handshake’s hello command, the connection POOL must clear close all connections with the matching
+:code:`serviceId`.
+
+
+**Implementation Notes**
+------------------------
+
+All connections MUST track a generation number. For non-load balanced deployments, the connection’s
```

Comment:
As I mentioned in the CMAP spec comment, this information about generation numbers should probably go somewhere else. Not sure where all of it goes, though.

Reply:
Have noted this in the CMAP spec now.